WIP: osd,librados: add dmclock capabilities to librados allowing clients to specify QoS

sseshasa opened the pull request:
This is WIP PR to incorporate changes to enable dmclock capability to librados
thereby allowing clients to specify QoS. The commits in this PR are pulled from
an earlier attempt as part of https://github.com/ceph/ceph/pull/20235 that was
not merged.

The original commits are modified suitably to work with the current
"mclock_scheduler" implementation. The original authors are credited in the
adapted commits.

Unit tests have been written to exercise the functionality added thus far.
Additional ideas to test (with an actual Rados client using the new APIs) are
being explored.

Update (09/06/2022):
Additional tests were performed after incorporating the new librados
APIs into the Fio's source code. The QoS params set by Fio client were
confirmed to reach the mClock queue through logs.

Signed-off-by: Sridhar Seshasayee <sseshasa@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

sseshasa commented:
jenkins test make check

sseshasa commented:
jenkins test make check arm64

sseshasa commented:
jenkins test make check

sseshasa commented:
jenkins test classic perf

sseshasa commented:
jenkins retest this please

mattbenjamin commented:
@sseshasa if you're building on another author's work, that author should be appropriately credited, not merely by reference to the unmerged PR.

sseshasa commented:
> @sseshasa if you're building on another author's work, that author should be appropriately credited, not merely by reference to the unmerged PR. To be clear, I think that issue needs to be resolved before this or any similar PR could be merged.

Absolutely, I will make it clearer in the PR description. The adapted commits credits the original authors and the PR it was adapted from. I am open to your suggestions on this.

mattbenjamin commented:
> > @sseshasa if you're building on another author's work, that author should be appropriately credited, not merely by reference to the unmerged PR. To be clear, I think that issue needs to be resolved before this or any similar PR could be merged.
> 
> Absolutely, I will make it clearer in the PR description. The adapted commits credits the original authors and the PR it was adapted from. I am open to your suggestions on this.

If it's not convenient to keep the original commits, building on those in new commits (sometimes it isn't), then the primary commit message should indicate the original authors and summarize what their contribution was.  It's not sufficient to do this in the PR description alone, as that goes away when the PR is merged.

jdurgin commented:
> > > @sseshasa if you're building on another author's work, that author should be appropriately credited, not merely by reference to the unmerged PR. To be clear, I think that issue needs to be resolved before this or any similar PR could be merged.
> > 
> > 
> > Absolutely, I will make it clearer in the PR description. The adapted commits credits the original authors and the PR it was adapted from. I am open to your suggestions on this.
> 
> If it's not convenient to keep the original commits, building on those in new commits (sometimes it isn't), then the primary commit message should indicate the original authors and summarize what their contribution was. It's not sufficient to do this in the PR description alone, as that goes away when the PR is merged.

It looks like that's already included in the commit messages, is there something else you're looking for?

mattbenjamin commented:
> > > > @sseshasa if you're building on another author's work, that author should be appropriately credited, not merely by reference to the unmerged PR. To be clear, I think that issue needs to be resolved before this or any similar PR could be merged.
> > > 
> > > 
> > > Absolutely, I will make it clearer in the PR description. The adapted commits credits the original authors and the PR it was adapted from. I am open to your suggestions on this.
> > 
> > 
> > If it's not convenient to keep the original commits, building on those in new commits (sometimes it isn't), then the primary commit message should indicate the original authors and summarize what their contribution was. It's not sufficient to do this in the PR description alone, as that goes away when the PR is merged.
> 
> It looks like that's already included in the commit messages, is there something else you're looking for?

Yes.  I found this:
`NOTE: The original changes are from commit
https://github.com/ceph/ceph/commit/c27fb6ef437ae749232261fb1ff6bb3a1a112006 from PR:
https://github.com/ceph/ceph/pull/20235.`
The intent is clearly there.  What I'm asking for is for the author and a description of the changes to be inserted, as text, here.  The links to unmerged git commits are potentially not stable, I feel.  We should just say who did what, right here.
Matt

sseshasa commented:
> > > > > @sseshasa if you're building on another author's work, that author should be appropriately credited, not merely by reference to the unmerged PR. To be clear, I think that issue needs to be resolved before this or any similar PR could be merged.
> > > > 
> > > > 
> > > > Absolutely, I will make it clearer in the PR description. The adapted commits credits the original authors and the PR it was adapted from. I am open to your suggestions on this.
> > > 
> > > 
> > > If it's not convenient to keep the original commits, building on those in new commits (sometimes it isn't), then the primary commit message should indicate the original authors and summarize what their contribution was. It's not sufficient to do this in the PR description alone, as that goes away when the PR is merged.
> > 
> > 
> > It looks like that's already included in the commit messages, is there something else you're looking for?
> 
> Yes. I found this: `NOTE: The original changes are from commit https://github.com/ceph/ceph/commit/c27fb6ef437ae749232261fb1ff6bb3a1a112006 from PR: https://github.com/ceph/ceph/pull/20235.` The intent is clearly there. What I'm asking for is for the author and a description of the changes to be inserted, as text, here. The links to unmerged git commits are potentially not stable, I feel. We should just say who did what, right here. Matt

@mattbenjamin Understood. I will append a summary text that mentions what the original authors did in the original commit and a section to describe what was adapted from that commit. I can make additional changes based on your comments going forward. Thanks.

ronen-fr commented on src/common/mClockCommon.h:
```diff
@@ -0,0 +1,59 @@
+
+namespace ceph {
+
+namespace dmc = crimson::dmclock;
```
crimson?

sseshasa commented:
> > 
> > Yes. I found this: `NOTE: The original changes are from commit https://github.com/ceph/ceph/commit/c27fb6ef437ae749232261fb1ff6bb3a1a112006 from PR: https://github.com/ceph/ceph/pull/20235.` The intent is clearly there. What I'm asking for is for the author and a description of the changes to be inserted, as text, here. The links to unmerged git commits are potentially not stable, I feel. We should just say who did what, right here. Matt
> 
> @mattbenjamin Understood. I will append a summary text that mentions what the original authors did in the original commit and a section to describe what was adapted from that commit. I can make additional changes based on your comments going forward. Thanks.

@mattbenjamin It's been a while since I got to this PR. I have modified the relevant commit messages to convey:
- The "Change description" (i.e. what's adapted) section,
- The commit(s) from where they were adapted from,
- The original commit, description and author(s) are mentioned. ,
- Author who adapted the changes. I added an "Adapted by" section, and finally,
- The "signed-off-by" credits the original author(s).

Does this look agreeable to you?

sseshasa commented:
> @mattbenjamin It's been a while since I got to this PR. I have modified the relevant commit messages to convey:
> 
>     * The "Change description" (i.e. what's adapted) section,
> 
>     * The commit(s) from where they were adapted from,
> 
>     * The original commit, description and author(s) are mentioned. ,
> 
>     * Author who adapted the changes. I added an "Adapted by" section, and finally,
> 
>     * The "signed-off-by" credits the original author(s).
> 
> 
> Does this look agreeable to you?

@mattbenjamin Could you please take a look? Thanks!

sseshasa commented:
@athanatos @ivancich @jdurgin @cbodley @neha-ojha Can you please start reviewing this? At this point only the librados APIs have been written and tested. I will continue to add changes involving Ceph options that clients can use to set the QoS params among other minor changes. But this shouldn't prevent the reviews from progressing. Thanks!

sseshasa commented:
jenkins test make check

sseshasa commented:
jenkins test make check

athanatos commented on src/messages/MOSDOp.h:
```diff
@@ -7,9 +7,9 @@
  * Foundation.  See file COPYING.
- * 
+ *
  */
```
Unnecessary whitespace changes

sseshasa replied:
I guess these came in from the commits of original PR. These are removed now.

athanatos commented on src/messages/MOSDOpReply.h:
```diff
@@ -47,6 +48,8 @@ class MOSDOpReply final : public Message {
   bool do_redirect;
   request_redirect_t redirect;
+  dmc::PhaseType qos_phase;
+  uint64_t qos_cost;
```
uint64_t qos_cost = 1u; here rather than in the default constructor below, easier to audit.

sseshasa replied:
Done.

athanatos commented on src/messages/MOSDOpReply.h:
```diff
@@ -216,15 +224,27 @@ class MOSDOpReply final : public Message {
       encode_trace(payload, features);
     }
   }
+
```
Unnecessary whitespace change.

sseshasa replied:
Removed.

athanatos commented on src/osd/osd_types.h:
```diff
@@ -6606,4 +6606,56 @@ using missing_map_t = std::map<hobject_t,
+  uint64_t qos_profile_id;
+
+  client_qos_params_t() :
+    reservation(1), weight(1), limit(999999), qos_profile_id(0)
```
Set these defaults with the declaration.  limit should probably be std::numeric_limits<uint64_t>::max();

athanatos replied:
```
  uint64_t reservation = 1;
  uint64_t weight = 1;
  uint64_t limit = std::numeric_limits<uin64_t>::max();
...
  client_qos_params_t() = default;
```

sseshasa replied:
Done.

athanatos commented on src/osd/scheduler/OpSchedulerItem.h:
```diff
@@ -156,6 +169,16 @@ class OpSchedulerItem {
   uint64_t get_owner() const { return owner; }
   epoch_t get_map_epoch() const { return map_epoch; }
+  dmc::ReqParams get_qos_req_params() const { return qos_req_params; }
+  void set_qos_req_params(dmc::ReqParams qparams) {
```
Does this actually have a user?  Looks like the constructor is the only place this gets set.

sseshasa replied:
Currently there are no users and therefore I removed it for now.

athanatos commented on src/osd/scheduler/OpSchedulerItem.h:
```diff
@@ -156,6 +169,16 @@ class OpSchedulerItem {
+  const client_qos_params_t& get_qos_profile_params() const {
+    return qos_profile_params;
+  }
+  void set_profile_params(client_qos_params_t qparams) {
```
This also doesn't appear to have a user.

sseshasa replied:
Removed for now as there is no user.

athanatos commented on src/osd/scheduler/mClockScheduler.cc:
```diff
@@ -76,15 +77,18 @@ void mClockScheduler::ClientRegistry::update_from_config(const ConfigProxy &conf
-  else
-    return &(ret->second);
+  }
+  return &(ret->second);
```
```
} else {
  return &(ret->second);
}
```

sseshasa replied:
Done.

athanatos commented on src/osd/scheduler/mClockScheduler.cc:
```diff
@@ -76,15 +77,18 @@ void mClockScheduler::ClientRegistry::update_from_config(const ConfigProxy &conf
 const dmc::ClientInfo *mClockScheduler::ClientRegistry::get_external_client(
   const client_profile_id_t &client) const
 {
+  // rw_mtx must be held by the Caller
```
Let's replace this comment with an assert that rw_mtx is held.

sseshasa replied:
The rw_mtx is now replaced with a ceph::mutex and the comment is replaced with assertion using ceph_mutex_is_locked() macro.

athanatos commented:
I made some superficial comments, still working on a proper review.

athanatos commented on src/common/mClockCommon.h:
```diff
@@ -0,0 +1,59 @@
+  uint8_t int_phase;
+  decode((uint8_t&)int_phase, p);
+  phase = static_cast<dmc::PhaseType>(int_phase);
+}
```
We should probably create denc instances for these (see include/denc.h) since they are fixed size.  Is there a reason we need to define these here rather than where the types are actually defined?

athanatos replied:
These implementation choose not to include versioning.  That's probably ok, but was it intentional?

sseshasa replied:
Looking into this.

athanatos replied:
On reflection, it's probably not worth versioning these structures.  It would add overhead to MOSDOp[Reply] encoding, and if we really need to we can use the versioning on MOSDOp[Reply].

It would probably be worth replacing these encode/decode implementations with denc instances, though.  Because the structs are defined elsewhere, you'll need to implement the denc overloads and denc_traits<> template directly.

sseshasa replied:
@athanatos The encode/decode implementation is now replaced with denc_traits<> template. Please verify.

sseshasa replied:
@athanatos Please review the latest changes that incorporates denc_traits when you get a chance. Thanks!

sseshasa replied:
@athanatos Can you please review the mClockCommon.h changes? The changes incorporate the denc_traits as per your suggestion.

athanatos replied:
That part LGTM

athanatos commented on src/osd/scheduler/OpSchedulerItem.h:
```diff
@@ -123,7 +126,17 @@ class OpSchedulerItem {
       map_epoch(e)
-  { qos_cost = 0; qos_item = false; }
+  {
+    qos_cost = 0; qos_item = false;
```
These need to be on separate lines.

sseshasa replied:
Done.

athanatos commented on src/osd/scheduler/mClockScheduler.h:
```diff
@@ -113,15 +119,90 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
+        cleaning_job = std::unique_ptr<crimson::RunEvery>(
+          new crimson::RunEvery(get_clear_every(),
+                                std::bind(&ClientRegistry::do_clean,
+                                this)));
```
RunEvery looks like another implementation of SafeTimer.  Rather than auditing RunEvery, I'd really rather we just stick to SafeTimer.

sseshasa replied:
Makes sense. I have re-implemented the same functionality with SafeTimer. I had to accordingly replace the std::shared_mutex with ceph::mutex.

athanatos commented on src/osdc/QosProfileMgr.cc:
```diff
@@ -0,0 +1,57 @@
+                                                  uint64_t wgt,
+                                                  uint64_t lim)
+{
+  return create(res, wgt, lim, next_client_profile_id++);
```
next_client_profile_id isn't protected by the lock here -- you probably need to create a create_helper() so that the two create() overloads can take the lock and call it.

sseshasa replied:
Yes, that's correct! create_qos_profile() is now called by the two create() overloads after acquiring the lock.

athanatos commented on src/osd/scheduler/mClockScheduler.h:
```diff
@@ -34,9 +34,14 @@ namespace ceph::osd::scheduler {
 
 constexpr uint64_t default_min = 1;
 constexpr uint64_t default_max = 999999;
+constexpr auto default_clear_every = std::chrono::seconds(60);
```
default_clear_period

sseshasa replied:
Changed.

athanatos commented on src/osd/scheduler/mClockScheduler.h:
```diff
@@ -113,15 +119,90 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
+    using MarkPoint = std::pair<TimePoint, counter_t>;
+    std::deque<MarkPoint> clear_mark_points;
+    // Frequency of the clean-up job
+    Duration clear_every = default_clear_every;
```
clear_period

sseshasa replied:
Changed.

athanatos commented on src/osd/scheduler/mClockScheduler.h:
```diff
@@ -113,15 +119,90 @@ class mClockScheduler : public OpScheduler, md_config_obs_t {
+    using TimePoint = decltype(std::chrono::steady_clock::now());
+    using MarkPoint = std::pair<TimePoint, counter_t>;
+    std::deque<MarkPoint> clear_mark_points;
+    // Frequency of the clean-up job
```
// Period between runs of do_clean()

sseshasa replied:
Changed.

athanatos commented:
I can't think of any tests in the rados thrash suite that would exercise the background cleaning thread.

athanatos commented:
Ok, I'm done with this pass, re-request a review from me when you'd like me to take another look.

athanatos reviewed (changes requested):
See above comments.

sseshasa commented:
> I can't think of any tests in the rados thrash suite that would exercise the background cleaning thread.

Currently there are no tests. I was able to modify fio (the rados engine) to use the new librados APIs to create QoSProfile and perform IOs on a rados pool. I am looking into adding a rados_qos engine in fio and if that could later be used by teuthology.

sseshasa commented:
jenkins test make check

sseshasa commented:
jenkins test make check

sseshasa commented:
jenkins test api

sseshasa commented:
jenkins test make check

rzarzynski commented:
Note from PR scrub: this one is going to updated soon.

cbodley commented:
> the rgw team is still very interested in the use case i presented [on the original PR](https://github.com/ceph/ceph/pull/20235#issuecomment-383192689), which i vaguely recall being represented with negative client ids but lost track of where that happened in the code
> 
> we'd love to help with testing of that use case when you feel it's ready

we're still very interested in this! as implemented, is it possible to create a profile that's shared between multiple librados clients (rgw instances)?

we'd really like a way to separate rgw's client i/o from our background i/o like garbage collection, but we'd need for those 'client' and 'background' profiles to be shared instead of per-rgw

cc @ivancich @mattbenjamin

mattbenjamin commented:
Thanks @cbodley .  I would like to understand better how we achieved a shared QoS domain between cooperating RGWs in the original PR.

> 
> we'd really like a way to separate rgw's client i/o from our background i/o like garbage collection, but we'd need for those 'client' and 'background' profiles to be shared instead of per-rgw
> 
> cc @ivancich @mattbenjamin

cbodley commented:
> Thanks @cbodley . I would like to understand better how we achieved a shared QoS domain between cooperating RGWs in the original PR.

what i recall from discussions with @ivancich was the idea to use negative profile IDs to represent these shared mclock classes, but it doesn't look like this was ever implemented. there would be some open questions about how to map these negative IDs to ceph daemons and their mclock configurations

it's likely that we'd want something more general though, where librados clients can create/modify/fetch these shared profiles, with their distributed state managed somewhere like ceph-mon

sseshasa commented:
jenkins test api

sseshasa commented:
jenkins test make check

sseshasa commented:
Re-opening this.
